private/bufpkg/bufimage/bufimageutil/image_index.go

```diff
@@ -39,13 +39,22 @@ type imageIndex struct {
 }
 
 // newImageIndexForImage builds an imageIndex for a given image.
-func newImageIndexForImage(image bufimage.Image) (*imageIndex, error) {
+func newImageIndexForImage(image bufimage.Image, opts *imageFilterOptions) (*imageIndex, error) {
 	index := &imageIndex{
 		NameToDescriptor: make(map[string]protosource.NamedDescriptor),
-		NameToExtensions: make(map[string][]protosource.Field),
-		NameToOptions:    make(map[string]map[int32]protosource.Field),
 	}
+	if opts.includeCustomOptions {
+		index.NameToOptions = make(map[string]map[int32]protosource.Field)
+	}
+	if opts.includeKnownExtensions {
+		index.NameToExtensions = make(map[string][]protosource.Field)
+	}
+
 	for _, file := range image.Files() {
+		// TODO: protosource.File is a heavyweight representation, and NewFile is not a cheap
```

Comment:
In this case, a simple pass over the hierarchy (like already done to build the image index) is the most efficient. We just need to add tracking of fully-qualified names during that walk and then also add an extra map to the index for upward links. That way, we don't need to do anything else and can use the descriptor protos directly.

Both old `protoreflect` (mine) and new (API v2) are easier to work with since no index is needed. But they are definitely not lighter weight or more efficient. The real reason I see this as heavy-weight is because it has to build a complete mirror model of the entire descriptor hierarchy _in addition_ to the index. The advantage of `protoreflect` over `protosource` is that it has those indexes built in... but they aren't definitely not cheaper to construct. They are almost certainly more expensive because of the way it does file-at-a-time indexing/linking to create the descriptors and the fact that a second level of indexing is then needed to produce a `protoregistry.Types` (needed to do extension/custom option lookup).

You can see where I had implemented all of this logic on top of the API v2 `protoreflect` stuff [here](https://github.com/bufbuild/core/pull/6727/files#diff-f91fdd79fdbfbb2db1d93857dcc87000bc7a22691085e03ef088da3f7ae79030R55).

But I think we could do better in terms of making this lean and fast (which I think is a useful bit of code to improve and try to optimize if we expect it to be used more thanks to new `GetSchema` API endpoint).